rfd/0091-session-streaming.md

```diff
@@ -0,0 +1,238 @@
+---
+authors: Zac Bergquist <zac@goteleport.com>
+state: draft
+---
+
+# RFD 91 - Streaming Session Playback
+
+## Required Approvers
+
+* Engineering: @kimlisa && (@ibeckermayer || @ryanclark)
+* Product: @xinding33
+
+## What
+
+Update Teleport's session playback to use the `StreamSessionEvents` API.
+
+Teleport's session recordings are stored as an ordered stream of gzipped and
+protobuf-encoded events. See [RFD 2](./0002-streaming.md) for details.
+
+Prior versions of Teleport stored recordings in a different format where the
+session content (referred to as "chunks") was stored separately from the 
+metadata and timing information about those chunks.
+
+When RFD 2 was implemented and session streaming was added, the players
+(both the web UI and `tsh play`) saw minimal changes and we opted to convert
+the new protobuf event stream into the legacy metadata/chunks format for 
+playback.
+
+This RFD outlines the steps necessary to remove this legacy conversion step
+and support streaming for session playback directly.
+
+## Why
+
+There are several reasons why we'd like to make this change.
+
+### Removal of Legacy Code
+
+The `lib/events` package has a long history of [issues](https://github.com/gravitational/teleport/issues?q=is%3Aopen+is%3Aissue+label%3Aaudit-log)
+and dead code that
+[should](https://github.com/gravitational/teleport/pull/13217) 
+[be](https://github.com/gravitational/teleport/pull/12395) 
+[removed](https://github.com/gravitational/teleport/pull/11380).
+
+While we've succeeded in removing a large amount of old code, there is still
+a non-trivial amount of code left behind due to the legacy playback format
+that can be removed once we support the newer streaming API.
+
+Reducing the surface area of the packages related to the audit log and session
+recordings will make it easier to understand and modify as we continue to work
+on new efforts such as an encrypted audit log.
+
+### User Experience
+
+The required conversion to the old playback format results in an experience that
+is suboptimal for users. Since this process requires the entire session be 
+downloaded and converted before playback can begin, users experience significant
+latency when attempting to play back a large session.
```

Comment:
One thing noted in [RFD 0048](https://github.com/gravitational/teleport/blob/zmb3/web-session-stream/rfd/0048-desktop-access-session-recording.md#playback) is

> For SSH, this format has the nice property of the "chunks" file being useful on its own, as the raw data can be viewed and searched independent of the timing data (outside of the Teleport player).

This property will be lost with this change as the chunks file will no longer be generated or sent. It may be worthwhile figuring out how useful this property was to users and if it was seeing a lot of use, some way to still allow them to get this type of file (or another similar type).

Reply:
Yes, but that property was only useful when the chunks were raw text that you could grep/cat. Now that the chunks are protobuf, having the chunks file doesn't help you because you'd have to write special code to parse it.

See also
- https://github.com/gravitational/teleport/issues/11694
- https://github.com/gravitational/teleport/issues/4930

Reply:
@zmb3  Existing process does not require to download entire session, the chunks could be downloaded by demand and removed from memory once processed by term. The reason for UI to do that was something temporary that lasted longer than expected. 

> (too much memory usage, high latency, etc)

Just be careful with UX because "high latency"  was _never_ an issue but an ability to smoothly go back and forth is one of the key features used almost at every demo. So while code consistency is important, just make sure that you are not prioritizing it over UX/Product. 

As for streaming in general, I suggest to look into recordings (mp4) as that would give you the best UX for RDP sessions (with streaming it will never work nicely).